### PR TITLE
Deepspeech fixes

### PR DESCRIPTION
**Pull request by chandramouli-sastry**
This PR contains the following:
* Fixes device placement errors and corrections to the Feedforward module and BN to match jax implementation. 
* The deepspeech BN code extends SyncBatchNorm to use syncing only if in ddp. 
* Zeros out the padded-elements in jax deepspeech code to match pytorch code. 
* Adds a guard to sync the state-dicts only for conformers in DDP mode.

### CODE REVIEW

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py [hunk 1]
```diff
@@ -554,7 +554,7 @@ def __call__(
                 deterministic=deterministic)
         outputs = jnp.concatenate([outputs, backward_outputs], axis=-1)
         cell_idx += 1
-
+      outputs = outputs * (1 - input_paddings[:, :, None])
```
Comment (sourabh2k15) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:557]: why this change in jax deepspeech model ?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 2]
```diff
@@ -60,8 +60,10 @@ def init_model_fn(
     model.to(DEVICE)
     if N_GPUS > 1:
       if USE_PYTORCH_DDP:
+        self.requires_sync_before_eval = True
         model = DDP(model, device_ids=[RANK], output_device=RANK)
       else:
+        self.requires_sync_before_eval = False
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:66]: I think you should add this False line to before `if N_GPUS > 1:`

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py [hunk 3]
```diff
@@ -217,8 +223,13 @@ def reset_parameters(self) -> None:
       init.zeros_(self.weight)
       init.zeros_(self.bias)
 
-  def forward(self, input):  # pylint: disable=redefined-builtin
+  def forward(self, input):
+    # currently only GPU input is supported: added check below
+    # if not input.is_cuda:
+    #     raise ValueError("SyncBatchNorm expected input tensor to be on GPU")
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py:229]: delete or uncomment?

**Review by znado — COMMENTED** [review]
LGTM pending answering Sourabh's question

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 1 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:557]: Mainly because they help in easily matching pytorch and jax outputs. They correspond to the padded out portions of the outputs and aren't used by any later layers.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 2 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py]: Yes! Thank you so much for catching this!!!

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py (hunk 3 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py:229]: I have left it in so that it is easier for people to diff between this implementation and the original implementation?

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py (hunk 1 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py:557]: can you modify pytorch code instead to remove this line ?

Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py]: I have removed this line! I will later modify the test script to only consider non-padded outputs.

**Comment by sourabh2k15** [comment]
can you fix presubmits ?  changes LGTM 
after presubmits pass I can merge it

**Review by sourabh2k15 — COMMENTED** [review]
can you revert the jax deepspeech model file completely ?
